### PR TITLE
fixes issue #1368 by adding stub#resolvesThis

### PR DESCRIPTION
**Pull request by HugoMuller**
#### How to verify - mandatory
1. Check out this branch (see github instructions below)
2. `npm install`
3. `npm test`

### CODE REVIEW

File: lib/sinon/default-behaviors.js [hunk 1]
```diff
@@ -181,6 +183,12 @@ module.exports = {
         return fake;
     },
 
+    resolvesThis: function resolvesThis(fake) {
+        fake.resolveThis = true;
+        fake.resolve = false;
+        fake.reject = false;
```
Comment (lucasfcosta) [inline lib/sinon/default-behaviors.js]: IMO we should keep this consistent with other behaviors, such as `rejects`, which also sets many other flags, as you can see [here](https://github.com/sinonjs/sinon/pull/1517/files#diff-f46d81a3e31504be425e201260dab281R165). Even though this is enough to make it work, I think that by keeping the whole API consistent when it comes to flagging fakes is extremely important.

**Comment by lucasfcosta** [comment]
Hi @HugoMuller, thanks for this awesome Pull Request!

I really liked your tests and your clean code. It works, indeed and it is an extremely important feature IMO. I just had a comment regarding the flags being set when the `resolvesThis` behavior is used.

Please let me know if you disagree with anything or if you have any further doubts.

I'm looking forward to being able to merging this as soon as everyone approves it.

Thanks again for your work 😄

File: lib/sinon/default-behaviors.js (hunk 1 above)
Comment (fatso83) [inline lib/sinon/default-behaviors.js]: I think you created the wrong link? It just points back to your comment.

File: test/stub-test.js [hunk 2]
```diff
@@ -285,6 +285,65 @@ describe("stub", function () {
...
+
+            return instance.stub().then(function (actual) {
+                assert.same(actual, instance);
+            });
+        });
+
+        var strictMode = (function () {
+            return this;
+        }()) === undefined;
+        if (strictMode) {
```
Comment (mroderick) [inline test/stub-test.js]: I am not sure a conditional for strict mode is necessary:

* Sinon's source is meant to be ES5.1, strict mode and expects to run in engines that supports that.
* The test file has `use strict;` at the top.

If for some reason, someone decides to run the tests in non-strict mode or in engines that doesn't support that, then I think the tests **should** be allowed to fail.

File: test/stub-test.js [hunk 3]
```diff
@@ -285,6 +285,65 @@ describe("stub", function () {
...
+                stub.resolvesThis();
+
+                // Due to strict mode, would be `global` otherwise
+                return stub().then(function (actual) {
+                    assert.same(actual, undefined);
+                });
+            });
+        }
+
+        it("stub respects call", function () {
```
Comment (mroderick) [inline test/stub-test.js]: Can this expectation be more specific, to make it easier to understand what "respects call" means?

File: test/stub-test.js [hunk 4]
```diff
@@ -285,6 +285,65 @@ describe("stub", function () {
...
+            var stub = createStub.create();
+            stub.resolvesThis();
+            var object = {};
+
+            return stub.call(object).then(function (actual) {
+                assert.same(actual, object);
+            });
+        });
+
+        it("stub respects apply", function () {
```
Comment (mroderick) [inline test/stub-test.js]: Can this expectation be more specific, to make it easier to understand what "respects apply" means?

File: test/stub-test.js [hunk 5]
```diff
@@ -2171,6 +2230,16 @@ describe("stub", function () {
             refute.defined(instance.stub());
         });
 
+        it("cleans 'resolvesThis' behavior", function () {
```
Comment (mroderick) [inline test/stub-test.js]: It is generally better to be very specific when writing expectations, especially in codebases with distributed contributors. Generic expectations often lead to readers trying to derive the intent from the implementation of the test. If there's a mistake in the implementation, then the reader is likely to end up with an incorrect understanding of then intended behaviour.

I would suggest rewriting the expectation to something like this:

```js
it("cleans 'resolvesThis' behavior using stub.resetBehavior", function () {
it("calls stub's resetBehavior method to clean behavior", function () {
```

File: test/stub-test.js (hunk 2 above)
Comment (HugoMuller) [inline test/stub-test.js]: Actually, the tests I wrote are very similar to those written for returnsThis.
So the same logic applies here. But if this one does not make sense, then I can remove it.
Same remark for the unclear expectations.

File: test/stub-test.js (hunk 5 above)
Comment (HugoMuller) [inline test/stub-test.js]: I'll fix these expectations. Thanks for the suggestions.

File: lib/sinon/default-behaviors.js (hunk 1 above)
Comment (HugoMuller) [inline lib/sinon/default-behaviors.js]: I guess I can set `returnValueDefined` and `exception` flags to false to keep consistency. Do you think some other flags need to be turned off?

**Comment by mroderick** [comment]
👍

File: lib/sinon/default-behaviors.js (hunk 1 above)
Comment (lucasfcosta) [inline lib/sinon/default-behaviors.js]: @fatso83 I think that link only works inside the files tab. When I click it inside of it the link works.

Basically I think the following flags should be set:

* `fake.returnValue`
* `fake.returnValueDefined`
* `fake.exception`

Comment (HugoMuller) [inline lib/sinon/default-behaviors.js]: Okay now resolves, rejects and resolvesThis all act on the same flags.
I set returnValue to undefined as the stub resolves with itself.
Is it okay by you?

**Pull request merged**